JP-4000: Persistence Options Added

kmacdonald-stsci opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Resolves [JP-4000](https://jira.stsci.edu/browse/JP-4000)

<!-- describe the changes comprising this PR here -->
This PR eliminates the old persistence step code and implementing a new step for using the persistence flag as an informational flag that is based on saturation events. This flagging persists across integrations and can persist across exposures using an output file. The interface for this step has changed, as well.

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [ ] If you have a specific reviewer in mind, tag them.
- [x] add a build milestone, i.e. `Build 12.0` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see [changelog readme](https://github.com/spacetelescope/jwst/blob/main/changes/README.rst) for instructions) 
  - [x] update or add relevant tests
  - [x] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [x] Do truth files need to be updated ("okified")?
      - [x] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [x] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

kmacdonald-stsci committed:
Updating the pipeline changing the location of the persistence step.

kmacdonald-stsci committed:
Updating detector1 pipeline.

kmacdonald-stsci committed:
Adding persistence flagging parameters.

kmacdonald-stsci committed:
Updating step tests.

kmacdonald-stsci committed:
Adding persistence window handler.

kmacdonald-stsci committed:
Updating how the persistence step handles the persistence options.

kmacdonald-stsci committed:
Updating the persistence flagging handler.

kmacdonald-stsci committed:
Corrected misspelling.

kmacdonald-stsci committed:
Adding some CI tests to test the new persistence flagging.

tapastro commented on jwst/persistence/tests/test_persistence_step.py:
```diff
@@ -1,13 +1,51 @@
+    assert np.array_equal(res.groupdq, model.groupdq)
+
+
+def test_persistence_time_10_sec(create_sci_model):
```
Rename to 70 sec?

kmacdonald-stsci committed:
Changing type for persistence_array to list.

kmacdonald-stsci committed:
Added tests for persistence flagging.

kmacdonald-stsci committed:
Adding comments for development.

kmacdonald-stsci committed:
Use ASDF for persistence flagging information array.

kmacdonald-stsci committed:
Update tesets to handle using ASDF for persistence array.

kmacdonald-stsci committed:
Updating the pipeline call for the persistence step.

kmacdonald-stsci committed:
Updating the ASDF file name for outputting new persistence array.

kmacdonald-stsci committed:
Moving persistence step to original place in pipeline to test regression tests.

kmacdonald-stsci committed:
Updating the persistence array filename.

kmacdonald-stsci committed:
Changing pipeline from regression testing purposes to correct placement for this PR. Old pipeline is commented out and should be removed before this PR is approved.

kmacdonald-stsci committed:
Updating the pipeline to run the persistence step.

kmacdonald-stsci committed:
Updating the way integration times are computed.

kmacdonald-stsci committed:
Updating persistence array output file name creation.

kmacdonald-stsci committed:
Added test stub for test that needs to be developed to use int_times, rather than doing timing math using the exposure start time.

kmacdonald-stsci committed:
Adding comment to check the correct start time of the exposure if the int_times table doesn't exist.

kmacdonald-stsci committed:
Removing unneeded tests.

kmacdonald-stsci committed:
Removing old, unwanted code and updating the tests.

kmacdonald-stsci committed:
Adding an epoch_to_mjd function.

kmacdonald-stsci committed:
Updating testing for persistence.

kmacdonald-stsci committed:
Updating the use of int_times to compute the current_time for persistence flagging.

kmacdonald-stsci committed:
Updating the computation of epoch_time.

kmacdonald-stsci committed:
Adding int_times testing and updating a test to use the correct exposure start time.

kmacdonald-stsci committed:
The trapdensity is no longer used, so is removed from the regression test. I expect this to cause failures in the regression tests.

kmacdonald-stsci committed:
Updating return values for method, removing no longer used variables.

kmacdonald-stsci committed:
Updating the way the output filename for the persistence arrays gets created.

kmacdonald-stsci committed:
Updating CI tests for saving the persistence array.

kmacdonald-stsci committed:
Updating the documentation for the persistence step to match the rewrite of the persistence step. Reference files are no longer used, so that has been removed from the documentation.

melanieclarke reviewed (commented):
In addition to addressing David's comments from testing, this needs some preliminary clean up on the software side before we continue reviews:
1. pre-commit, changelog, and docs tests are all failing.
2. unit tests are failing (some notes below on how to fix)
3. the regression test runs only a trivial use case.  We'll need regression tests that demonstrate both: initial persistence flagging and saving the asdf file, and using an existing asdf file to flag persistence in a subsequent exposure.

drlaw1558 commented:
Adding a couple other quick comments:
- Oftentimes it will be necessary to pass persistence information between *multiple* exposures.  In the example I'm working with for instance, persistence in program 5014 affects later exposure in 5014 and also in the next program PID 5407.  Does the save_persistence infrastructure pass along input persistence information from the prior exposure (input via the asdf file) in the output asdf file, or just from the exposure being processed?  It seems like it might be easiest to have a single reference asdf file that gets added to each time a new exposure is processed rather than many such files?  There's the risk of collision of course, but since the file would be an agglomeration of info from many exposures anyway perhaps that doesn't matter, allowing the filename to be more predictable.
- It may be worth default to some finite-valued persistence_time instead of None.  Users can always customize it based on their needs, but a starting point might be informative so long as the results are only affecting the DQ plane.  This would also deal with the issue that the step might run, and report completion, but be doing nothing if this keyword were not set.  Just as a stand-in for now I'd say perhaps 900 seconds, and instruments can always customize further with param ref files.

kmacdonald-stsci committed:
Updating the use of the pixel dictionary for dqflags to the group dictionary for the persistence flag.

kmacdonald-stsci committed:
Beginning to update persistence regression test.

kmacdonald-stsci committed:
Updating change log.

tapastro commented:
> The output persistence file is then titled jw02772101001_02101_00001_nrcalong_pers20260318113812109525_pers20260318122231819751.asdf. That should be fixed to not concatenate the pers-timestamp, but to replace it.

This issue is still present - we need some filename cleaning of the input persistence array filename if it is also requested that the step save a fresh version. I'd also suggest that we truncate the timestamp by ~3 digits - we probably don't need to go beyond milliseconds.

kmacdonald-stsci committed:
Updating the file nameing convention for output persistence array file.

kmacdonald-stsci committed:
Returned the location of the persistence array file to be saved in the current working directory.

kmacdonald-stsci commented:
> > The output persistence file is then titled jw02772101001_02101_00001_nrcalong_pers20260318113812109525_pers20260318122231819751.asdf. That should be fixed to not concatenate the pers-timestamp, but to replace it.
> 
> This issue is still present - we need some filename cleaning of the input persistence array filename if it is also requested that the step save a fresh version. I'd also suggest that we truncate the timestamp by ~3 digits - we probably don't need to go beyond milliseconds.

This has now been fixed. Also, I reduced the time string by using on the last two digits of the year and removing the microseconds.

kmacdonald-stsci committed:
Updating test for saving persistence array file in output directory.

kmacdonald-stsci committed:
Adding functionality to put output persistence array file in the pipeline output directory.

tapastro commented:
Sorry for asking for a fix to your fix, but after testing the updates more (and reading David's requests above), we should modify how we save the persistence array file.

The dynamically generated filename with the timestamp does not lend itself well to programmatic use; we expect some users will want to generate an array file for the first exposure in a list, then reuse that file to store the evolving state of persistence as exposures are processed. 

Let's modify the `save_persistence_array` keyword to be a string rather than a bool. If provided (i.e. not None), use that as the output filename.

kmacdonald-stsci committed:
Updating how to save the persistence file array and the corresponding test.

kmacdonald-stsci committed:
Updating the argument description for the persistence step.

kmacdonald-stsci committed:
Updating the persistence step description to add the new functionality of the step.

kmacdonald-stsci committed:
Updating the spec for the step on the proper use of a parameter.

tapastro commented on docs/jwst/persistence/arguments.rst:
```diff
@@ -1,26 +1,36 @@
+This boolean determines if the ``DO_NOT_USE`` flag will get set when the ``PERSISTENCE``
+flag gets set.
+
+*  ``--skip``
```
You can drop this parameter - it's a default parameter for all steps and is typically not documented in the step-specific documentation.

tapastro replied:
@kmacdonald-stsci pinging you only because it sounds like you were done with this, but I think this still needs addressing.

kmacdonald-stsci replied:
Yes. I removed it from the code, but forgot the docs. Thanks! I'll fix it this afternoon.

tapastro commented on docs/jwst/persistence/arguments.rst:
```diff
@@ -1,26 +1,36 @@
+        persistence_dnu = boolean(default=False) # If True the set the DO_NOT_USE flag with PERSISTENCE
+        skip = boolean(default=False) # Skip the persistence step entirely
+
 The persistence step has three step-specific arguments.
```
I think this should be four rather than three.

kmacdonald-stsci committed:
Updating due to code review feedback.

kmacdonald-stsci committed:
Removing unneeded documentation.

drlaw1558 commented:
Thanks @kmacdonald-stsci ; a few comments from testing this again.

With the test data set jw05014008002_02107_00001_mirimage_uncal.fits I'm still seeing some unexpected behavior interfacing with the ramp fitting step.  In particular, having a GROUPDQ value of 32 is causing ramp_fit to return a zero slope.  In integration 2, (X,Y) = (697,482) the original DQ values are [1, 1, 0, 0, 4, 2, 2, 2, 2, 3] and ramp_fit gives a reasonable slope.  After the persistence step runs the DQ values are [33, 33, 32, 32, 36, 34, 34, 34, 34, 35] which is what I'd expect, with the persistence flag being set but otherwise no DQ changes since persistence_dnu=False
ramp_fit though is turning this into a zero slope (not NaN, zero).  I think we'll have to figure out what's going on in ramp_fit and address that at the same time.

Other comments:

- I was talking with a user having issues with persistence from pixels that nearly saturate, but never quite reach the saturation level.  One potential solution to that would be to add an optional parameter to the persistence step to add additional flagging over some raw DN threshold.  This obviously wouldn't be perfect, as the ramp zeropoint has been subtracted already, but it could nonetheless be a useful knob for some programs to dial in an offline capacity.  Do you think it would be practical to add an option 'custom_dn_threshold' (or something like that) to treat any pixel for which the SCI array was larger than that value as if the SATURATED bit had been set for the purposes of setting PERSISTENCE flags?  Your thoughts also welcome here @tapastro 

- I'm not sure I'm understanding the persistence array file correctly.  Running on the first exposure in a sequence I can set save_persistence='myfile.asdf'.  Running on the second exposure I can set persistence_array_file='myfile.asdf' to use that input to help flag things.  I'm not sure how the time information contained in the persistance_data array is used though; it's time for the persistence window but where is the zeropoint for where that time began?  It also looks like I can't append information to the same persistence file; if I try to do so I think the new information must be overwriting the old as the array is still 1024x1032?  Appending seems necessary if the persistence time is long enough that the persistence from one exposure might impact multiple later exposures, and I could thus build up a comprehensive persistence file that the step would determine from timestamps which parts of it were relevant to use at any given time.

kmacdonald-stsci commented:
> Thanks @kmacdonald-stsci ; a few comments from testing this again.
> 
> With the test data set jw05014008002_02107_00001_mirimage_uncal.fits I'm still seeing some unexpected behavior interfacing with the ramp fitting step. In particular, having a GROUPDQ value of 32 is causing ramp_fit to return a zero slope. In integration 2, (X,Y) = (697,482) the original DQ values are [1, 1, 0, 0, 4, 2, 2, 2, 2, 3] and ramp_fit gives a reasonable slope. After the persistence step runs the DQ values are [33, 33, 32, 32, 36, 34, 34, 34, 34, 35] which is what I'd expect, with the persistence flag being set but otherwise no DQ changes since persistence_dnu=False ramp_fit though is turning this into a zero slope (not NaN, zero). I think we'll have to figure out what's going on in ramp_fit and address that at the same time.
> 
> Other comments:
> 
> * I was talking with a user having issues with persistence from pixels that nearly saturate, but never quite reach the saturation level.  One potential solution to that would be to add an optional parameter to the persistence step to add additional flagging over some raw DN threshold.  This obviously wouldn't be perfect, as the ramp zeropoint has been subtracted already, but it could nonetheless be a useful knob for some programs to dial in an offline capacity.  Do you think it would be practical to add an option 'custom_dn_threshold' (or something like that) to treat any pixel for which the SCI array was larger than that value as if the SATURATED bit had been set for the purposes of setting PERSISTENCE flags?  Your thoughts also welcome here @tapastro
> * I'm not sure I'm understanding the persistence array file correctly.  Running on the first exposure in a sequence I can set save_persistence='myfile.asdf'.  Running on the second exposure I can set persistence_array_file='myfile.asdf' to use that input to help flag things.  I'm not sure how the time information contained in the persistance_data array is used though; it's time for the persistence window but where is the zeropoint for where that time began?  It also looks like I can't append information to the same persistence file; if I try to do so I think the new information must be overwriting the old as the array is still 1024x1032?  Appending seems necessary if the persistence time is long enough that the persistence from one exposure might impact multiple later exposures, and I could thus build up a comprehensive persistence file that the step would determine from timestamps which parts of it were relevant to use at any given time.

I'll take a look at the pixel (x, y) = (697, 482).

Tyler will have to reply to the first of your other comments.

With regard to what the persistence array file is:

The way the persistence array is used is that if pers_arr[row, col] is zero, no flagging occurs. If it is non-zero, it is an epoch time that defines the end a persistence window. A zero point for the beginning is not necessary because a zero value assumes the pixel is currently not in a persistence window. Once a pixel is determined to be in a pixel window, the start time is not necessary, only the end time. To clarify, the logic is as follows:

- Compute the start time of each integration
https://github.com/kmacdonald-stsci/jwst/blob/34af6d6e6ce7231262c144dd245753ba5d0a202b/jwst/persistence/persistence.py#L106

- For each group, add the group time to the current_time, to give the time for each group
https://github.com/kmacdonald-stsci/jwst/blob/34af6d6e6ce7231262c144dd245753ba5d0a202b/jwst/persistence/persistence.py#L112

- Since the pers_arr contains the end time of a persistence window, if the current time is later than the time in pers_arr[row, col], then the (row, col) is no longer in the persistence window, so no flagging, hence pers_arr[row, col] gets set to zero
https://github.com/kmacdonald-stsci/jwst/blob/34af6d6e6ce7231262c144dd245753ba5d0a202b/jwst/persistence/persistence.py#L149

- If the current group for pixel (row, col) is determined to be the first saturation flag of a pixel ramp, then pers_arr[row, col] is set to current_time + persistence_time.
https://github.com/kmacdonald-stsci/jwst/blob/34af6d6e6ce7231262c144dd245753ba5d0a202b/jwst/persistence/persistence.py#L157

- Any pixel in pers_arr that is non-zero gets the current group flagged with the persistence flag (and the do not use flag, if that option is selected).
https://github.com/kmacdonald-stsci/jwst/blob/34af6d6e6ce7231262c144dd245753ba5d0a202b/jwst/persistence/persistence.py#L165

After processing the last group in the last integration, whatever is in the pers_arr is what gets saved to the path defined for save_persistence ("myfile.asdf"). The saved array is a list of each pixel. Those with zero entries are not in any persistence window. Those with non-zero entries (of time in epoch time) will still be in persistence windows that close after the last group in the last integration. 

If that is passed in as persistence_array_file, it will be a 2D array file with zero entries for pixels not in the last defined persistence widnow, whereas non-zero entries will be the epoch time of the end of the current persistence window. This allows the persistence window to carry over from the last group of the last integration of the previously processed RampModel to the first group of the first integration of the current RampModel being processed.

drlaw1558 commented:
@kmacdonald-stsci Ah; it looks like the problem was on my end.  In my testing I wasn't correctly folding in https://github.com/spacetelescope/stcal/pull/538
Thanks for the clarification on how the epoch times work; I'll retest with the right stcal this time.

drlaw1558 commented:
Ok, I've repeated my tests using the correct version of stcal.
- Ignore what I said in my previous comment about unexpected behavior in the ramp fitting; that looks like it's behaving as expected when using the correct stcal version.  rateints are unchanged when only the PERSISTENCE flag is set, and NaNs in the expected locations when persistence_dnu=True
- I see now how the information is being written to the persistence asdf file and confirm that it can both be passed into processing for a given file and written out again modified with new persistence pixels.  Flagging locations between multiple exposures in a sequence look as expected.

Since this is a feature primarily for use by external users (likely in a notebook environment) though, I think the asdf file format needs to be hardened against the misuse that it's almost certain to get.  Since the file currently records only an end time for the persistence, it's possible to propagate the flagging backwards in time.  Say you reduce your data once, allowing exposures 1-4 to add information to the asdf file.  When you rerun the notebook the asdf file already exists, and since the persistence caused by exposure 4 has an end time later than exposure 1, exposure 1 will be flagged according to persistence that will be produced by exposure 4.  Arguably you should always re-create the file each time the pipeline is run, and never run out of exposure-order sequence, but I'm certain many users won't do that correctly once they think the persistence info file has been built.

I think what it needs is to specify both the start and end times for persistence to avoid the backward-in-time flagging, possibly with a new plane for each exposure processed (since a pixel might saturate and affect exposures 1-3, then saturate again and affect exposures 6-8).  It seems pretty expensive to record this in 2d array format, particularly if we start looking at planes for each exposure.  Maybe it would thus be easier to save the information as a long table of the form (filename, x, y, tstart, tstop)?  For a fairly typical single MIRI exposure with about 5000 saturating pixels this would be a 5x5000 array compared to a 2000x2000 array, which could be appended to easily, do a filename check to avoid appending the same information multiple times, and bound an event both forward and backward in time.  Thoughts?

kmacdonald-stsci commented:
@dlaw The situation you're describing sounds like someone who fundamentally doesn't understand the usage of persistence. Such misuse of tools cannot be defended against.

I understand your concern about backwards flagging, but what to do in that situation? If a user runs an earlier exposure using a persistence array of a later exposure and the current time occurs before the beginning of the persistence window, this should generate a processing exceptions, as the processing has clearly be used incorrectly, with a clear error message.

If a user will misuse it this way, the user will misuse it in other ways. For example, say the user uses a persistence array file that was created using a persistence window of 200 seconds, then used this file later with a persistence window of 300 seconds, this will obviously have inconsistencies.

We should have expect a reasonable level of competence from users. That said, we can do some basic parameter checks that raise exceptions. A simple proposal could be something like the following.

To save space, instead of a 2D array with a beginning and end time, since the persistence window is uniform, I can add the persistence window time. When setting up the tree at the link below, I could add the `"persistence_time" : persistence_time` to the dictionary. During processing I could add an additional check to see if the current time is earlier than the end time minus the `persistence_time`, indicating a backwards in time flagging. As I said above, I think this should cause an exception. Further, when processing starts, the inputted `persistence_time` will be checked against that in the file and if there is a difference, an exception will be raised. Also, as a part of the tree, a list of files processed could be in the persistence array file. Something like `"processed_files" : [list of files]`. This will never be empty as the ramp file used to create the persistence array file will be in this list. When processing, the current ramp file can be checked against this list, if it's in there, just return from processing logging an "already processed message". If it's not in there, append the ramp file to the persistence file list, then process it. Possibly, also list the computed time of the last group in the last integration. This time could be used as a check against the first group of the first integration of next processed ramp file, with an exception raised if the first group of the first integration occurs before the last group of the last integration in the last ramp processed. So the tree will look something like this:

```
tree = {
    "processed_files" : [file list],
    "persistence_time" : persistence_time,
    "persistence_data" : self.persistence_array,
    "last_group_time" : last_group_time,
}
```

Also, since the file list can be appended to and the `last_group_time` changes, any inputted persistence array file should always be rewritten, meaning that if `persistence_array_file` is used, then `save_persistence` must also be used. 

https://github.com/kmacdonald-stsci/jwst/blob/a917e71bea99d25ef28417bec1fd295a6d20c5fb/jwst/persistence/persistence_step.py#L116

kmacdonald-stsci committed:
Point STCAL dependency to correct branch.

tapastro commented:
I like that the existing file is detector-shaped - it allows for quick viewing and comparison against the data, which I think is helpful. 

We could include some metadata in the persistence array file that could hold either the exposure end time of the last exposure or a list of exposure end times that contributed to the file - a comparison to that could raise an exception if the current process is trying to apply the file to an exposure prior to the existing times.

kmacdonald-stsci committed:
Skipping tests while updating persistence file array.

kmacdonald-stsci committed:
Updating how the persistence array file is structured in order to prevent backward in time flagging, i.e., erroneously flagging before the time window begins.

kmacdonald-stsci committed:
Adding testing of backwards flagging, but needs to be more thoroughly tested.

kmacdonald-stsci committed:
In progress commit for testing.

kmacdonald-stsci committed:
Cleaned up code and added new tests to property test the persistence array file, guarding against mismatched persistence times, as well as guarding against backwards flagging.

kmacdonald-stsci committed:
Updating based on ruff checks.

drlaw1558 commented on jwst/persistence/persistence.py:
```diff
@@ -214,857 +89,135 @@ def do_all(self):
+            for group in range(ngroups):
+                current_time = current_time + group_time
+                if self.persistence_time is not None:
+                    self.process_persistence_flagging(current_time, sat_array, integ, group)
```
I think we need
sat_array[:]=0
after this loop over groups and before continuing to the next integration.

I'm working with some very long test data with 120 groups, 8 ints that saturates on ~ group 60, using a persistence time of 3600 seconds.  At present, the ending persistence time is 3785 seconds after the start of the exposure.  That's not flagging nearly enough though, as persistence will also be caused by the 8th integration nearly 3600 seconds after the first integration.  With the modification above the ending persistence time recorded in the asdf file is 6133 seconds, which is more accurately going to flag an hour of data after the last integration on which saturation was reached.

kmacdonald-stsci replied:
I made this change.

kmacdonald-stsci committed:
Update style.

kmacdonald-stsci committed:
Removing reference to files no longer used.

kmacdonald-stsci committed:
Updating style code checks.

kmacdonald-stsci committed:
Update style checks.

kmacdonald-stsci committed:
Updating code style checks.

kmacdonald-stsci committed:
Setting sat_array to zero for each integration.

kmacdonald-stsci committed:
Added metadata needed to run CI tests with the new persistence steps.

tapastro reviewed (approved):
Thanks for your efforts, Ken! I think we're good to go.

tapastro commented:
Scratch that - I'll get the stcal PR in, then we'll need to update the pyproject here to point back to stcal/main.

I'm also going to get one more set of RTs going, just to have it in hand.

RTs here, with both PRs: https://github.com/spacetelescope/RegressionTests/actions/runs/27361455924

tapastro reviewed (dismissed):
RT run shows errors caused by existing regtests that use parameters which no longer exist: see test_niriss_image and test_nircam_persistence. Those tests will need to be updated to be compliant with the new step spec before we can merge.

kmacdonald-stsci committed:
Removed bad arguments causing errors.

kmacdonald-stsci committed:
Updating step option for persistence step regression test.

kmacdonald-stsci commented:
> RT run shows errors caused by existing regtests that use parameters which no longer exist: see test_niriss_image and test_nircam_persistence. Those tests will need to be updated to be compliant with the new step spec before we can merge.

The tests have been updated and no fails with "unrecognized arguments" occur.

https://github.com/spacetelescope/RegressionTests/actions/runs/27393127467/job/80954796120

tapastro commented on jwst/regtest/test_nircam_persistence.py:
This test is failing due to a missing output file. Your update changed the behavior of the test to only process one file rather than two - the current test here tests the second exposure against a truth value of that second exposure, while the new update only processes the first exposure. The simplest solution may just be to update your fixture above to run on the second exposure - that avoids the need to upload a new truth file.

kmacdonald-stsci replied:
I added the processing of the second file back to the test.

kmacdonald-stsci committed:
Adding back accidentally removed second file processing from test.

melanieclarke commented on pyproject.toml:
```diff
@@ -31,9 +31,8 @@ dependencies = [
-    # "stdatamodels>=4.1.0,<4.2.0",
-    "stcal @ git+https://github.com/spacetelescope/stcal.git@main",
     # "stcal>=1.18,<2",
+    "stcal @ git+https://github.com/kmacdonald-stsci/stcal.git@jp_4000_persistence_02",
```
Just leaving this comment as a reminder that this change needs to be reverted before merging.

melanieclarke replied:
The stcal PR is merged, so this can be updated now.

melanieclarke commented:
New regression test run:
https://github.com/spacetelescope/RegressionTests/actions/runs/27550654413

melanieclarke commented:
Looking through the last regression test run, I see NIRSpec detector1 products now have S_PERSIS = "COMPLETE" (e.g. test_nirspec_irs2_detector1[rate]: jwst.regtest.test_nirspec_irs2_detector1). 

This is unexpected to me - I didn't think we would be running persistence by default for any mode.  @hayescr @drlaw1558 - Does this need a parameter file update for NIRSpec?

kmacdonald-stsci commented:
> Looking through the last regression test run, I see NIRSpec detector1 products now have S_PERSIS = "COMPLETE" (e.g. test_nirspec_irs2_detector1[rate]: jwst.regtest.test_nirspec_irs2_detector1).
> 
> This is unexpected to me - I didn't think we would be running persistence by default for any mode. @hayescr @drlaw1558 - Does this need a parameter file update for NIRSpec?

This is likely due to the change in `calwebb_detector1.py‎`, which changes where this step gets run, which is just before ramp fitting, and is now run unconditionally.

hayescr commented:
@kmacdonald-stsci does the default running of the new persistence step impact the output rate images from detector1?  If it is just impacting qualitative data flags, that's not necessarily a problem to have on by default.  If we are changing the resulting rates though, we might want to deliver a parameter file update to turn this off by default (as is currently done) until we've had more of a chance to test the changes.  I thought I saw earlier comments that the persistence flagging might impact ramp fitting, is that still the case?

Tagging @Rplesha for awareness.

Rplesha commented:
Perhaps instead we can set the default to be off in the code? If everyone is delivering a parameter reference file to turn it off, it seems like we want folks to opt into it for now. Just a thought!

kmacdonald-stsci commented:
> @kmacdonald-stsci does the default running of the new persistence step impact the output rate images from detector1? If it is just impacting qualitative data flags, that's not necessarily a problem to have on by default. If we are changing the resulting rates though, we might want to deliver a parameter file update to turn this off by default (as is currently done) until we've had more of a chance to test the changes. I thought I saw earlier comments that the persistence flagging might impact ramp fitting, is that still the case?
> 
> Tagging @Rplesha for awareness.

Persistence flags do not impact ramp fitting. There is a corresponding PR in STCAL that handles persistence flags to prevent it from impacting ramp fitting.

drlaw1558 commented:
It should be the case that the persistence step is currently doing nothing; the step runs but with a default persistence time of None I wouldn't expect the step to have any impact on even QA flags until such time as param ref files are made that set values explicitly.  As such the persistence step is *effectively* skipping even though it says otherwise.

It might be worth making that more obvious.  A clear thing to do would be to add a log statement to the effect of 'No persistence time information found, so not doing anything'.  Arguably the step status should be Failed in that case too, because it tried to compute persistence but was unable to do so.  Having a 'Failed' step status in all of our data products probably isn't desirable though for sake of concern/questions it might cause to end users.

I think that talks me around to having the persistence step currently skip by default in this build; not because of any worry about what it does to the data (as it does nothing) but to avoid confusion over the fact that it's currently doing nothing.

In the next build I'd like to get some param ref files in place to set default persistence timescale for each instrument, and those can enable the step in inform-only mode.

tapastro commented:
This sounds reasonable to me. By the book, I agree that the current scenario should probably populate the status with FAILED - I could see a case for populating with SKIPPED if no persistence time is provided, but it does go a bit against the update to allow FAILED as a status if the step is attempted.

kmacdonald-stsci committed:
Adding FAILED status if persistence_time is not valid.

kmacdonald-stsci commented:
> This sounds reasonable to me. By the book, I agree that the current scenario should probably populate the status with FAILED - I could see a case for populating with SKIPPED if no persistence time is provided, but it does go a bit against the update to allow FAILED as a status if the step is attempted.

I just added a condition to set as FAILED with an invalid persistence_time.

tapastro commented:
> I just added a condition to set as FAILED with an invalid persistence_time.

That sounds fine, but as David stated above, we don't want to cause alarm if our default outputs in MAST are showing that a step FAILED. We should also set the skip parameter to have a default value of True in the step spec; we can let instruments turn the step on as they deliver parameter reference files with defined persistence times. 

We can revisit the default skip status of the step if/when all instruments have specified their desired behavior through ref files.

kmacdonald-stsci commented:
> > I just added a condition to set as FAILED with an invalid persistence_time.
> 
> That sounds fine, but as David stated above, we don't want to cause alarm if our default outputs in MAST are showing that a step FAILED. We should also set the skip parameter to have a default value of True in the step spec; we can let instruments turn the step on as they deliver parameter reference files with defined persistence times.
> 
> We can revisit the default skip status of the step if/when all instruments have specified their desired behavior through ref files.

I added a `skipped` parameter to the step that defaults to `True` and skips the test.

tapastro commented on docs/jwst/persistence/arguments.rst:
```diff
@@ -1,26 +1,35 @@
+This boolean determines if the ``DO_NOT_USE`` flag will get set when the ``PERSISTENCE``
+flag gets set.
+
+*  ``--skipped``
```
There is a default parameter for all steps, `skip` - no need to add a new one.

tapastro commented on jwst/persistence/persistence_step.py:
```diff
@@ -16,15 +19,13 @@ class PersistenceStep(Step):
+        persistence_time = integer(default=None) # Time, in seconds, to use for persistence window
+        persistence_array_file = string(default=None) # A path to an ASDF file containing a 2-D array of persistence times per pixel
+        persistence_dnu = boolean(default=False) # If True the set the DO_NOT_USE flag with PERSISTENCE
+        skipped = boolean(default=True) # Default step to be skipped.
```
`skipped` -> `skip`

kmacdonald-stsci replied:
I `git reset` the changes.

kmacdonald-stsci replied:
I reverted this out of the spec. The parameter `skipped` (nor `skip`) is no longer part of the spec. 

I used `skipped` because `skip` didn't work as a default `False`. I suspect the parent class overrides this.

melanieclarke replied:
You need to set `skip = boolean(default=True)` in the spec to override the parent class default and skip the step by default.  It should look exactly like the example Tyler posted above -- it's used like this in many pipeline steps.

tapastro reviewed (commented):
There is a default param to use, see e.g. https://github.com/spacetelescope/jwst/blob/2d18ee436be3448a18f0e0279bd7d48e12512a8f/jwst/adaptive_trace_model/adaptive_trace_model_step.py#L25

for an example. Sorry I wasn't explicit about that.

kmacdonald-stsci committed:
Updating tests to check for FAILED status.

kmacdonald-stsci committed:
Updating build dependency for STCAL.

kmacdonald-stsci committed:
Updating pyproject.toml

kmacdonald-stsci committed:
Updating STCAL dependency.

kmacdonald-stsci committed:
Adding  to spec.

melanieclarke committed:
Clean up change note and pyproject.toml

melanieclarke commented:
I pushed up some minor clean up for the change note and pyproject.toml, but I think this is otherwise ready for merge now.  I verified locally that NIRSpec detector1 now skips persistence, as expected.

When RTs are clean again, I will run another set of regression tests for okify, then get this merged.

melanieclarke commented:
RTs for okify:
https://github.com/spacetelescope/RegressionTests/actions/runs/27632447310

Diffs are as expected: persistence is skipped for all modes and there are expected changes to the persistence regtest.
